Support for parameters in prepared statements

rschlussel opened the pull request:
@electrum

losipiuk commented on presto-parser/src/main/java/com/facebook/presto/sql/tree/RewritingTraversalVisitor.java:
```diff
@@ -0,0 +1,491 @@
+
+        QueryBody queryBody = (QueryBody) process(node.getQueryBody(), context);
+
+        List<SortItem> orderBy = ImmutableList.copyOf(Lists.transform(node.getOrderBy(), (value -> (SortItem) process(value, context))));
```
We probably want to use java streams for transform.
Sth like:
```java
node.getOrderBy().stream().map(value -> (SortItem) process(value, context)).collect(toImmutableList())
```

It will probably require you to `com.facebook.presto.util.ImmutableCollectors` from `presto-main` to `presto-parser` due to lack of dependency.

rschlussel replied:
ok.  I'll change that throughout

rschlussel commented:
updated to use streams (squashed it into the original RewritingTraversalVisitor commit)

electrum commented:
Can you split the "Implement RewritingTraversalVisitor" commit to do the "private -> protected" change separately? It touches a lot of files and makes the other commit harder to read.

electrum commented on presto-parser/src/main/java/com/facebook/presto/sql/tree/RewritingTraversalVisitor.java:
```diff
@@ -0,0 +1,490 @@
+
+import static com.facebook.presto.sql.util.ImmutableCollectors.toImmutableList;
+
+public abstract class RewritingTraversalVisitor<C>
```
I think `RewritingVisitor` is a better name, since "traversal" doesn't seem to make sense here

electrum replied:
This class seems to duplicate much of `ExpressionTreeRewriter`. Is there a reason it can't reuse that?

electrum commented on presto-parser/src/main/java/com/facebook/presto/sql/util/ImmutableCollectors.java:
```diff
@@ -0,0 +1,89 @@
+import java.util.function.Function;
+import java.util.stream.Collector;
+
+public final class ImmutableCollectors
```
Remove the unused methods

electrum commented on presto-parser/src/main/java/com/facebook/presto/sql/tree/RewritingTraversalVisitor.java:
```diff
@@ -0,0 +1,490 @@
+    @Override
+    protected Node visitNode(Node node, C context)
+    {
+        return node;
```
This should probably throw, just like `ExpressionTreeRewriter`, otherwise anything not handled could have the wrong behavior:
```java
throw new UnsupportedOperationException("not yet implemented: " + node.getClass().getName());

martint commented:
Why are we doing this as a tree rewrite? IMO, this should be done by the analyzer/planner. The AST should not be modified.

electrum commented on presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java:
```diff
@@ -1323,6 +1332,12 @@ public Node visitInterval(SqlBaseParser.IntervalContext context)
+    @Override
+    public Node visitParameter(SqlBaseParser.ParameterContext context)
+    {
+        return new Parameter(getLocation(context), parameterPosition++);
```
Split out the increment to be a separate statement

electrum commented on presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java:
```diff
@@ -346,7 +351,11 @@ public Node visitDeallocate(SqlBaseParser.DeallocateContext context)
     {
         String name = context.identifier().getText();
-        return new Execute(getLocation(context), name);
+        List<Literal> parameterValues = new ArrayList<>(context.literal().size());
```
I think this can be
```java
List<Literal> values = visit(context.literal(), Literal.class);

electrum commented on presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java:
```diff
@@ -350,8 +356,26 @@ public static Statement unwrapExecuteStatement(Statement statement, SqlParser sq
+        // validate that we have the right number of parameters
+        validateParameters(unwrapped, execute.getParameters());
+
+        // replace Parameter expressions with the supplied values.
```
Nit: don't end this in a period (it doesn't start with a capital so it isn't a sentence)

electrum commented on presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java:
```diff
@@ -350,8 +356,26 @@ public static Statement unwrapExecuteStatement(Statement statement, SqlParser sq
+        ParameterCollector collector = new ParameterCollector();
+        collector.process(node, null);
+        if (parameterValues.size() != collector.getParameterCount()) {
+            throw new SemanticException(INVALID_PARAMETER_USAGE, node, format("Incorrect number of parameters: expected %d but found %d", collector.getParameterCount(), parameterValues.size()));
```
`SemanticException` already takes format args

electrum replied:
Also, always `%s` instead of a plain `%d` since it does the same thing but is safer (won't throw if the argument isn't a number)

electrum commented on presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4:
```diff
@@ -292,6 +287,16 @@ primaryExpression
     | '(' expression ')'                                                             #parenthesizedExpression
     ;
 
+literal
```
This won't allow array literals (or other types of non-scalar literals)

electrum commented on presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4:
```diff
@@ -258,13 +258,8 @@ valueExpression
-    | STRING                                                                         #stringLiteral
-    | BINARY_LITERAL                                                                 #binaryLiteral
+    : literal                                                                        #literalExpression
+    | PARAMETER                                                                      #parameter
```
Just make this `'?'` rather than defining a new token

electrum commented on presto-parser/src/main/java/com/facebook/presto/sql/tree/Execute.java:
```diff
@@ -67,14 +77,15 @@ public boolean equals(Object obj)
         }
         Execute o = (Execute) obj;
-        return Objects.equals(name, o.name);
+        return Objects.equals(name, o.name) && Objects.equals(parameters, o.parameters);
```
Format like
```java
return Objects.equals(name, o.name) &&
        Objects.equals(parameters, o.parameters);

electrum commented on presto-parser/src/main/java/com/facebook/presto/sql/tree/Parameter.java:
```diff
@@ -0,0 +1,70 @@
+import java.util.Optional;
+
+public class Parameter
+    extends Expression
```
`extends` should be indented 8 spaces

electrum commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -6536,6 +6537,20 @@ public void testExecute() throws Exception
+    {
+        String query = "SELECT a + ?, count(1) FROM (VALUES 1, 2, 3, 2) t(a) GROUP BY a + ? HAVING count(1) > ?";
+        Session session = getSession().withPreparedStatement("my_query", query);
+        MaterializedResult actual = computeActual(session, "EXECUTE my_query USING 1, 1, 0");
```
Avoid creating a result by hand if possible:
```java
assertQuery(session,
        "EXECUTE my_query USING 1, 1, 0",
        "VALUES (2, 1), (3, 2), (4, 1)");

electrum commented on presto-parser/src/main/java/com/facebook/presto/sql/SqlFormatter.java:
```diff
@@ -174,6 +175,12 @@ protected Void visitExecute(Execute node, Integer indent)
+            List<Literal> parameters = node.getParameters();
+            if (!parameters.isEmpty()) {
+                builder.append(" USING ");
+                String parameterString = Joiner.on(", ").join(parameters);
```
You can do
```java
Joiner.on(", ").appendTo(builder, parameters);

electrum commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/SemanticErrorCode.java:
```diff
@@ -80,4 +80,6 @@
 
     AMBIGUOUS_FUNCTION_CALL,
+
+    INVALID_PARAMETER_USAGE
```
End with trailing comma

electrum commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -809,6 +811,12 @@ else if (valueType.equals(TIMESTAMP)) {
         @Override
+        protected Type visitParameter(Parameter node, StackableAstVisitorContext<AnalysisContext> context)
+        {
+            throw new SemanticException(INVALID_PARAMETER_USAGE, node, "Parameters are only allowed in prepared statements");
```
Add a test for this

electrum commented on presto-main/src/test/java/com/facebook/presto/execution/TestUnwrapExecute.java:
```diff
@@ -55,6 +60,46 @@ public void testExecuteStatement()
+            fail("expected exception");
+        }
+        catch (SemanticException e) {
+            assertEquals(e.getCode(), INVALID_PARAMETER_USAGE);
```
Also validate the error message

electrum commented on presto-main/src/test/java/com/facebook/presto/execution/TestUnwrapExecute.java:
```diff
@@ -55,6 +60,46 @@ public void testExecuteStatement()
+            fail("expected exception");
+        }
+        catch (SemanticException e) {
+            assertEquals(e.getCode(), INVALID_PARAMETER_USAGE);
```
Here, too

rschlussel commented:
@martint replacing parameters is different from general query rewrites because until you do that, you don't really have the query at all.  You basically just have a template.  I know that in general we don't modify the AST in Presto, and query rewrites are done with PlanNodes, but this is more about getting the AST correct to begin with.   

If you don't replace the parameters until planning/analysis, you would have to add special casing everywhere and in general make the code much more complicated. For example, if you haven't replaced your Parameters, you couldn't test data type coercions appropriately.  In addition, you would have to thread the parameter values into the analyzer/planner because the replacement values come in through the execute statement, but by the time you're at analysis/planning your AST is the actual prepared statement.

rschlussel commented on presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4:
```diff
@@ -295,6 +295,8 @@ literal
     | booleanValue                                                                   #booleanLiteral
     | STRING                                                                         #stringLiteral
     | BINARY_LITERAL                                                                 #binaryLiteral
+    | ARRAY '[' (literal (',' literal)*)? ']'                                  #arrayLiteralConstructor
```
I don't like that the array/ row construction is duplicated, but I'm not sure how else to only allow literal arrays/rows without removing support for other expressions in arrays elsewhere.   It also still doesn't support maps, since those are created with functions.  Any ideas?

electrum replied:
Why not allow full expressions for parsing `EXECUTE`, but fail when replacing if the expression contains an identifier? This is similar to `CALL` or `SET SESSION`.

rschlussel replied:
I like that.  Commit has been updated

rschlussel commented:
@electrum addressed comments

rschlussel commented:
Fixed checkstyle violations.  Now it should be good.

martint commented:
> If you don't replace the parameters until planning/analysis, you would have to add special casing everywhere and in general make the code much more complicated. For example, if you haven't replaced your Parameters, you couldn't test data type coercions appropriately. 

It would only need to be handled in the expression analyzer and expression interpreter. Just like it can find out the types of qualified name references, it should be able to handle the "parameter" node to resolve its type and value. All that's needed is the analyzer to get the AST + the parameter->value map as input.

rschlussel commented:
@martint So basically you are suggesting:
1. pass the parameters to the DateDefinitionExecution and QueryExecution.
2. Change the ExpressionAnalyzer to gather the correct information about the parameters based on the values mapping
3. Add a rewriter for planning/optimization that rewrites the Parameters to the expressions that you want (because eventually you need to have the actual expression in the plan).
4. Change evaluateConstantExpression to also rewrite parameters to the actual values (used in SetSessionTask and CallTask, which don't go through the normal analysis/planning)

This still feels like you are handling something in three different places when the fundamental issue is that you don't really have your query yet.  Additionally,  it doesn't really make sense to me as a plan rewrite.  Since rewrites don't happen until after the initial plan has been created, you would come out of planning without a plan that can be executed.   I would have expected that any of the optimizers can be removed, and you would still have a valid plan (I don't know if we explicitly make that assumption, but it seems like it ought to be true, and the fact that we don't pass in arbitrary arguments to the optimizers implies to me that we expect a plan to be self-contained), but here you have essential execution information still sitting external to the plan.  To avoid this, you would have to rewrite an expression any time you try to add it to a plannode, which seems even worse.

rschlussel commented:
@martint I'm realizing that you may have intended the rewrite to happen during planning and not optimization.  I think that my point about handling parameters in several places instead of all at once still stands, but I'm not so familiar with the rewrites that happen in the QueryPlanner, so I'll take a look at those first.

rschlussel commented:
I looked at the QueryPlanner.  It's possible to add a rewrite at the end of planning instead of as an optimizer.   It would be the only rewrite that operates on the whole plan tree. Everything else in the planner operates in particular segments of  the plan.  But as far as I can see, a parameter rewrite couldn't do that because parameters can appear anywhere that you can have a literal expression so you need to either add a rewrite every time you could have an expression or you rewrite the whole plan at the end of planning.

In any case, putting a rewrite there solves the issue particular to having the rewrite in the optimizer, but doesn't resolve the sense of hacking together a solution in several places instead of doing it neatly in one place.  I still prefer the AST rewrite to properly fill out the query template before you start operating on it.

electrum commented:
@rschlussel, what @martint is saying is that this should happen during analysis in the `ExpressionAnalyzer`. Add a `visitParameter` that works similarly to `visitQualifiedNameReference`.

rschlussel commented:
@martint I updated the branch not to rewrite the AST, as you said.  But it's much more complicated than just adding the visitParameter() method to the ExpressionAnalyzer.  There are several reasons for this:
1. The analysis is called by so many things, and the parameters need to be threaded through to all of those places.  It's a pretty invasive change, though when you look at it, it's basically all the same places that you need to thread the metadata.  
2. As I had been saying earlier, the visitQualifiedNameReference method and analysis related changes take care of the analysis component, but the expressions still need to actually be rewritten for the PlanNodes. In the planner the QualifiedNameReferences get rewritten to symbols using the TranslationMap every time an expression is added to a PlanNode.   For parameters I did the same, and in all places where there were calls to rewrite the expression to symbols, I added rewrites to replace Parameters as well.
3.  Parameters can appear in both regular sql statements and data definition tasks.  In particular, SetSessionTask and CallTask need to replace parameters when they evaluate their arguments, and CreateViewTask needs to replace all the parameters when it calls getFormattedSql.  Additionally, the explains for DataDefinitionTasks need access to parameters.
4.  Parameters are never equal to each other, but the values that replace them can be.  This affects analysis of aggregations because you need to replace parameters there too so that you can accurately check whether a group by expression appears in the select list.

I still feel pretty strongly that the previous implementation was much neater.  You can find that branch here for your reference: https://github.com/Teradata/presto/compare/master...Teradata:parameters1

rschlussel commented:
@martint what do you think?

martint commented:
Sorry, I'm on vacation and haven't been following things much. I'll take a look when I get to a computer.

rschlussel commented:
no problem.  I didn't realize you were away.

rschlussel commented:
@martint ping

martint commented:
@rschlussel, I haven't forgotten :) It's on the list of things I'm looking at now.

rschlussel commented:
thanks :)

martint commented:
> But it's much more complicated than just adding the visitParameter() method to the ExpressionAnalyzer. There are several reasons for this:
> 1. The analysis is called by so many things, and the parameters need to be threaded through to all of those places. It's a pretty invasive change, though when you look at it, it's basically all the same places that you need to thread the metadata.
> 2. As I had been saying earlier, the visitQualifiedNameReference method and analysis related changes take care of the analysis component, but the expressions still need to actually be rewritten for the PlanNodes. In the planner the QualifiedNameReferences get rewritten to symbols using the TranslationMap every time an expression is added to a PlanNode. For parameters I did the same, and in all places where there were calls to rewrite the expression to symbols, I added rewrites to replace Parameters as well.

Yes, I agree that it has a bigger surface area *today*. But, consider that as part of the new optimizer work, we're going to decouple to SQL AST from the internal representation for scalar and relational expressions. So, all uses of ExpressionAnalyzer will go away (except for the query analyzer, of course).

The flow will look like this:

Text --(Parser)--> SQL AST --(Analyzer)--> SQL AST + annotations --(Planner)--> IR --(Optimizer)--> Optimizer IR

Analysis of parameter types will be done during query analysis, and replacing them with the corresponding values will be done during initial planning (which will be a simple desugaring and translation from SQL AST to IR). I expect `RelationPlanner` to be vastly simplified.


> 3. Parameters can appear in both regular sql statements and data definition tasks. In particular, SetSessionTask and CallTask need to replace parameters when they evaluate their arguments, and CreateViewTask needs to replace all the parameters when it calls getFormattedSql. Additionally, the explains for DataDefinitionTasks need access to parameters.

DataDefinitionTasks will eventually also be represented using the same IR as everything else and go through the same optimization and execution steps. Modeling them as a entirely different concept was one of those architectural detours we had to take for expediency, but it's not an ideal state of affairs.

> 4. Parameters are never equal to each other, but the values that replace them can be. This affects analysis of aggregations because you need to replace parameters there too so that you can accurately check whether a group by expression appears in the select list.

This is true, but if the parameters are semantically unrelated, it could be argued that the analysis should fail. This is similar to, say, selecting a column that's not in the group by clause but that happens to have a single constant value in the table.

One way out of this is to add support for named parameters, so users could refer to the same parameter in multiple positions in the query.

rschlussel commented:
I agree that the flow for the new optimizer much better.  I'll rebase this on master and resolve the merge conflicts, so it can be reviewed.

rschlussel commented:
rebased and ready for review again! 

I didn't change the group by analysis with parameters to fail, because to me it's the same as 
`select a +1, count(*) from table group by a+1`.  The expression in the select and the group by only happen to be the same, but we allow it.

rschlussel commented:
It's broken. I'm fixing it.

rschlussel commented:
fixed and rebased again.

martint commented on presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4:
```diff
@@ -78,7 +78,7 @@ statement
     | PREPARE identifier FROM statement                                #prepare
     | DEALLOCATE PREPARE identifier                                    #deallocate
-    | EXECUTE identifier                                               #execute
+    | EXECUTE identifier (USING expression (',' expression)*)?         #execute
```
Do we want to allow arbitrary expressions here or just literals?

rschlussel replied:
arbitrary expressions to allow for complex types.  We check in the analyzer that they are literals.

martint commented on presto-parser/src/main/java/com/facebook/presto/sql/tree/Execute.java:
```diff
@@ -23,28 +26,35 @@
     {
         super(location);
         this.name = requireNonNull(name, "name is null");
+        this.parameters = requireNonNull(ImmutableList.copyOf(parameters));
```
Add message to null check call

martint commented on presto-main/src/main/java/com/facebook/presto/execution/ParameterRewriter.java:
```diff
@@ -0,0 +1,80 @@
+    public Expression rewriteExpression(Expression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+    {
+        return treeRewriter.defaultRewrite(node, context);
+//        return coerceIfNecessary(node, rewrittenExpression);
```
remove commented-out code

martint commented on presto-main/src/main/java/com/facebook/presto/execution/ParameterRewriter.java:
```diff
@@ -0,0 +1,80 @@
+ * limitations under the License.
+ */
+
+package com.facebook.presto.execution;
```
This class should go either in `com.facebook.presto.sql.analyzer` or `com.facebook.presto.sql.planner`

martint commented on presto-main/src/main/java/com/facebook/presto/execution/ParameterCollector.java:
```diff
@@ -0,0 +1,39 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class ParameterCollector
```
Since this class is never used in a place where a visitor is needed, I'd restructure it to work like `DependencyExtractor`. I.e., make it a utility class with a static method named `extractParameters` or similar and hide the visitor-ness inside an inner class.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/AggregationAnalyzer.java:
```diff
@@ -81,20 +86,23 @@
         this.columnReferences = ImmutableSet.copyOf(columnReferences);
-        this.expressions = ImmutableList.copyOf(groupByExpressions);
+        this.parameters = parameters;
+        this.expressions = groupByExpressions.stream().map(e -> ExpressionTreeRewriter.rewriteWith(new ParameterRewriter(parameters), e)).collect(toImmutableList());
```
break it up into multiple lines for readability:

```java
this.expressions = groupByExpressions.stream()
        .map(...)
        .collect(...)
```

martint commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -790,7 +792,10 @@ else if (valueType.equals(TIMESTAMP)) {
         protected Type visitParameter(Parameter node, StackableAstVisitorContext<Void> context)
         {
-            throw new SemanticException(INVALID_PARAMETER_USAGE, node, "Parameters are only allowed in prepared statements");
+            checkArgument(node.getPosition() < parameters.size(), "invalid parameter number %s, max value is %s", node.getPosition(), parameters.size() - 1);
```
This should throw a SemanticException instead of IllegalArgumentException.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java:
```diff
@@ -243,7 +248,7 @@ public Object optimize(SymbolResolver inputs)
     }
 
-    private static class ConstantExpressionVerifierVisitor
+    public static class ConstantExpressionVerifierVisitor
```
Instead of exposing this class, add a method to ensure an expression is constant.

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/PlanBuilder.java:
```diff
@@ -27,18 +29,21 @@
     private final Optional<Symbol> sampleWeight;
 
-    public PlanBuilder(TranslationMap translations, PlanNode root, Optional<Symbol> sampleWeight)
+    public PlanBuilder(TranslationMap translations, PlanNode root, Optional<Symbol> sampleWeight, ParameterRewriter parameterRewriter)
```
Instead of taking a ParameterRewriter, this should just take the parameter values. It should construct a ParameterRewriter internally, as necessary, since that's an implementation detail. 

BTW, the Analysis object is available in the TranslationMap that's passed in (for the purpose of determining coercions -- arguably, this is something that should be passed separately.

martint commented on presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:
```diff
@@ -6965,6 +6982,17 @@ public void testExecute() throws Exception
+    public void testExecuteWithUsing()
+            throws Exception
+    {
+        String query = "SELECT a + ?, count(1) FROM (VALUES 1, 2, 3, 2) t(a) GROUP BY a + ? HAVING count(1) > ?";
```
I think this is going to be problematic in the long term. I expect we're going to want to analyze queries at preparation time in the future by allowing clients to provide hints about the types of the arguments. Allowing this will make it impossible, since there's no way to tell that the `?` in the select expression and the group by expression are the same.

As I commented before, I think a more robust way of supporting this kind of usage is to allow for named parameters. Something like:

```sql
SELECT a + :foo, count(1) FROM (VALUES 1, 2, 3, 2) t(a) GROUP BY a + :foo HAVING count(1) > ?";
```

martint commented:
Looks pretty good. Some minor comments and one point for further discussion. Back to you!

martint commented on presto-main/src/main/java/com/facebook/presto/execution/ParameterExtractor.java:
```diff
@@ -0,0 +1,52 @@
+    private static class ParameterExtractingVisitor
+            extends DefaultTraversalVisitor<Void, Void>
+    {
+        private final List<Parameter> parameters = new ArrayList<>();
```
This field is never used outside of this class. It seems to be used only for counting the number of parameters, so it could be replaced with a simple counter.

rschlussel replied:
Ah, yeah. It gets used later for describe input, but that's not in this pr.  I can change this to be a counter in the interim and change it back in the later commit.

rschlussel commented:
I discovered that you can't use parameters in a limit clause, since the parser only accepts "ALL" or an integer as valid limits.  Fixing this requires representing limits with an Expression instead of a String, which affects a number of places in the code.  I didn't want to address that here since this is already a very large PR, but I can do that as a follow on.

rschlussel commented:
I don't think the Travis failures were related to this change.  There's a passing build here: https://travis-ci.org/Teradata/presto/builds/153357471

rschlussel commented:
@martint when you have a chance

martint commented:
Yup, looking at it now. I was waiting for 0.152 release to go out.

rschlussel commented:
thanks

martint commented:
Looks good. Can you rebase and squash the fixups into their corresponding commits? I'll merge it then.

rschlussel commented:
squashed and rebased!

martint commented:
Merged, thanks!

dukehoops commented:
Hi, folks:

Congrats - eagerly anticipated. With which release do you think this
feature will get rolled out?

Thanks,

Nikita

On Aug 26, 2016 9:20 AM, "Martin Traverso" <notifications@github.com> wrote:

> Merged, thanks!
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/prestodb/presto/pull/5414#issuecomment-242781175>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAupuS1F2UkVOrBpMzD1Lcj_IhHYZb_tks5qjxIigaJpZM4IwSLT>
> .
>

rschlussel commented:
Thanks @martint!
@dukehoops this has been merged to master, so it will be available in the next release (0.153)
